File: sdks/python/speechswitch/websocket.py
"""Native asyncio WebSockets without third-party runtime dependencies.

RFC 6455 client framing: no extensions, compression, redirects or proxy discovery.
Providers own authentication, protocol codecs and connection creation.
"""

import asyncio
import base64
import hashlib
import os
import re
import ssl
import struct
from collections.abc import AsyncIterator, Mapping
from contextlib import asynccontextmanager
from typing import Protocol
from urllib.parse import quote, urlsplit


class WebSocketLike(Protocol):
    async def send(self, message: str | bytes) -> None:
        """Send one complete message, honoring cancellation and backpressure."""
        ...

    async def receive(self) -> str | bytes:
        """Read one message; normal close raises StopAsyncIteration.

        One reader at a time. Cancellation must release the connection.
        """
        ...

    async def aclose(self) -> None:
        """Release the connection and unblock pending reads/writes promptly."""
        ...


class WebSocketError(Exception):
    pass


class WebSocketClosed(WebSocketError):
    def __init__(self, code: int, reason: str = "") -> None:
        self.code = code
        self.reason = reason
        super().__init__(f"WebSocket closed ({code})")


def _frame(opcode: int, payload: bytes) -> bytes:
    size = len(payload)
    header = bytes([0x80 | opcode])
    if size < 126:
        header += bytes([0x80 | size])
    elif size < 65536:
        header += b"\xfe" + struct.pack("!H", size)
    else:
        header += b"\xff" + struct.pack("!Q", size)
    mask = os.urandom(4)
    return header + mask + bytes(byte ^ mask[index % 4] for index, byte in enumerate(payload))


class WebSocket:
    def __init__(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter, max_message_bytes: int) -> None:
        self._reader = reader
        self._writer = writer
        self._limit = max_message_bytes
        self._send_lock = asyncio.Lock()
        self._reading = False
        self._closed = False

    def _abort(self) -> None:
        self._closed = True
        self._writer.transport.abort()

    async def _send(self, opcode: int, payload: bytes) -> None:
        try:
            async with self._send_lock:
                if self._closed:
                    raise WebSocketClosed(1006)
                self._writer.write(_frame(opcode, payload))
                await self._writer.drain()
        except BaseException:
            self._abort()
            raise

    async def send(self, message: str | bytes) -> None:
        payload = message.encode("utf-8") if isinstance(message, str) else message
        if len(payload) > self._limit:
            raise WebSocketError("WebSocket message exceeds byte limit")
        await self._send(1 if isinstance(message, str) else 2, payload)

    async def receive(self) -> str | bytes:
        if self._reading:
            raise RuntimeError("WebSocket permits one reader at a time")
        if self._closed:
            raise StopAsyncIteration
        self._reading = True
        data = bytearray()
        message_opcode = 0
        try:
            while not self._closed:
                first, second = await self._reader.readexactly(2)
                final, opcode = bool(first & 0x80), first & 15
                if first & 0x70 or second & 0x80 or opcode not in (0, 1, 2, 8, 9, 10):
                    raise WebSocketError("Invalid WebSocket frame header")
                length = second & 127
                if opcode >= 8 and (not final or length > 125):
                    raise WebSocketError("Invalid WebSocket control frame")
                if length == 126:
                    length = struct.unpack("!H", await self._reader.readexactly(2))[0]
                    if length < 126:
                        raise WebSocketError("Nonminimal WebSocket frame length")
                elif length == 127:
                    length = struct.unpack("!Q", await self._reader.readexactly(8))[0]
                    if length < 65536 or length >= 1 << 63:
                        raise WebSocketError("Invalid WebSocket frame length")
                if opcode < 8:
                    if (opcode == 0) != (message_opcode != 0):
                        raise WebSocketError("Invalid WebSocket fragmentation")
                    if length > self._limit - len(data):
                        raise WebSocketError("WebSocket message exceeds byte limit")
                payload = await self._reader.readexactly(length)
                if opcode == 8:
                    if len(payload) == 1:
                        raise WebSocketError("Invalid WebSocket close payload")
                    code = struct.unpack("!H", payload[:2])[0] if payload else 1005
                    if payload and not (code in (1000, 1001, 1002, 1003, 1007, 1008, 1009, 1010, 1011, 1012, 1013, 1014) or 3000 <= code <= 4999):
                        raise WebSocketError("Invalid WebSocket close code")
                    reason = payload[2:].decode("utf-8")
                    await self._send(8, payload)
                    self._abort()
                    if code not in (1000, 1005):
                        raise WebSocketClosed(code, reason)
                    raise StopAsyncIteration
                if opcode == 9:
                    await self._send(10, payload)
                    continue
                if opcode == 10:
                    continue
                if opcode:
                    message_opcode = opcode
                data.extend(payload)
                if final:
                    return data.decode("utf-8") if message_opcode == 1 else bytes(data)
            raise StopAsyncIteration
        except asyncio.IncompleteReadError:
            self._abort()
            raise WebSocketClosed(1006) from None
        except UnicodeDecodeError:
            self._abort()
            raise WebSocketError("Invalid WebSocket UTF-8") from None
        except BaseException:
            self._abort()
            raise
        finally:
            self._reading = False

    async def aclose(self) -> None:
        if not self._closed:
            try:
                # Best effort close notification; cleanup never waits for a peer
                # handshake or a stuck transport drain. Cancellation aborts TCP.
                if not self._writer.is_closing():
                    self._writer.write(_frame(8, struct.pack("!H", 1000)))
            finally:
                self._abort()


@asynccontextmanager
async def connect_websocket(
    url: str,
    *,
    headers: Mapping[str, str] | None = None,
    ssl_context: ssl.SSLContext | None = None,
    max_message_bytes: int = 4 * 1024 * 1024,
    max_header_bytes: int = 64 * 1024,
) -> AsyncIterator[WebSocket]:
    """Own a direct ws/wss connection with certificate verification by default.

    Always use async with. A supplied SSLContext is an explicit trust override.
    Native authorization headers are supported; credentials are never redirected.
    """
    if type(max_message_bytes) is not int or max_message_bytes <= 0 or type(max_header_bytes) is not int or max_header_bytes <= 0:
        raise ValueError("WebSocket byte limits must be positive integers")
    try:
        target = urlsplit(url)
        host = target.hostname
        port = target.port if target.port is not None else (443 if target.scheme == "wss" else 80)
        if target.scheme not in ("ws", "wss") or not host or target.username is not None or target.password is not None or target.fragment or any(character.isspace() or ord(character) < 32 or character == "\\" for character in url):
            raise ValueError
        host = host.encode("idna").decode("ascii")
    except (ValueError, UnicodeError):
        raise ValueError("WebSocket URL must be ws/wss without credentials or a fragment") from None
    if target.scheme == "ws" and ssl_context is not None:
        raise ValueError("SSLContext requires a wss URL")
    authority = f"[{host}]" if ":" in host else host
    if target.port is not None:
        authority += f":{port}"
    path = quote(target.path or "/", safe="/%:@!$&'()*+,;=-._~")
    if target.query:
        path += "?" + quote(target.query, safe="/%?:@!$&'()*+,;=-._~")
    key = base64.b64encode(os.urandom(16)).decode("ascii")
    request_headers = {"Host": authority, "Upgrade": "websocket", "Connection": "Upgrade", "Sec-WebSocket-Key": key, "Sec-WebSocket-Version": "13"}
    protected = {name.lower() for name in request_headers} | {"sec-websocket-extensions", "sec-websocket-protocol", "content-length", "transfer-encoding"}
    for name, value in (headers or {}).items():
        if not re.fullmatch(r"[!#$%&'*+.^_`|~0-9A-Za-z-]+", name) or name.lower() in protected or any(ord(character) < 32 or ord(character) > 126 for character in value):
            raise ValueError("Invalid WebSocket request header")
        request_headers[name] = value
    handshake = (f"GET {path} HTTP/1.1\r\n" + "".join(f"{name}: {value}\r\n" for name, value in request_headers.items()) + "\r\n").encode("ascii")
    tls = (ssl_context or ssl.create_default_context()) if target.scheme == "wss" else None
    reader, writer = await asyncio.open_connection(host, port, ssl=tls, server_hostname=host if tls else None, limit=max_header_bytes)
    try:
        writer.write(handshake)
        await writer.drain()
        try:
            raw = await reader.readuntil(b"\r\n\r\n")
        except (asyncio.LimitOverrunError, asyncio.IncompleteReadError):
            raise WebSocketError("Invalid or oversized WebSocket handshake") from None
        if len(raw) > max_header_bytes:
            raise WebSocketError("Invalid or oversized WebSocket handshake")
        lines = raw[:-4].split(b"\r\n")
        status = lines[0].split(b" ", 2)
        if len(status) < 2 or status[0] != b"HTTP/1.1" or status[1] != b"101":
            raise WebSocketError("WebSocket handshake was not accepted")
        response: dict[bytes, list[bytes]] = {}
        for line in lines[1:]:
            name, separator, value = line.partition(b":")
            if not separator or not re.fullmatch(rb"[!#$%&'*+.^_`|~0-9A-Za-z-]+", name) or any(byte < 32 and byte != 9 or byte == 127 for byte in value):
                raise WebSocketError("Invalid WebSocket response header")
            response.setdefault(name.lower(), []).append(value.strip(b" \t"))
        upgrade = [token.strip().lower() for value in response.get(b"upgrade", []) for token in value.split(b",")]
        connection = [token.strip().lower() for value in response.get(b"connection", []) for token in value.split(b",")]
        expected = base64.b64encode(hashlib.sha1((key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11").encode("ascii"), usedforsecurity=False).digest())
        if upgrade != [b"websocket"] or b"upgrade" not in connection or response.get(b"sec-websocket-accept") != [expected] or b"sec-websocket-extensions" in response or b"sec-websocket-protocol" in response:
            raise WebSocketError("Invalid WebSocket upgrade response")
        socket = WebSocket(reader, writer, max_message_bytes)
        try:
            yield socket
        finally:
            await socket.aclose()
    except BaseException:
        writer.transport.abort()
        raise

File: sdks/go/runtime/websocket.go
package runtime

import (
	"context"
	"crypto/rand"
	"crypto/sha1"
	"crypto/tls"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"sync/atomic"
	"unicode/utf8"
)

type WebSocketMessage interface{ isWebSocketMessage() }
type WebSocketText string

func (WebSocketText) isWebSocketMessage() {}

type WebSocketBinary []byte

func (WebSocketBinary) isWebSocketMessage() {}

// WebSocketLike owns an already-connected socket. Methods must honor their
// context. Close is idempotent and unblocks pending sends/receives promptly.
type WebSocketLike interface {
	Send(context.Context, WebSocketMessage) error
	Receive(context.Context) (WebSocketMessage, error)
	Close() error
}

type WebSocketOptions struct {
	Header http.Header
	// Nil uses native net/http, verified TLS, HTTP/1.1, no redirects or proxies.
	// Overrides must preserve credential/cancellation ownership and return an
	// io.ReadWriteCloser body after upgrade. They own their raw header limits.
	Transport HTTPTransport
	// Zero selects 4 MiB per message and 64 KiB per native handshake.
	MaxMessageBytes int
	MaxHeaderBytes  int64
}

type WebSocketClosed struct {
	Code   int
	Reason string
}

func (e *WebSocketClosed) Error() string { return fmt.Sprintf("WebSocket closed (%d)", e.Code) }

// ConnectWebSocket opens RFC 6455 ws/wss without extensions or subprotocols.
// The provider constructs authentication at this boundary. Close aborts TCP;
// it never waits for a peer's graceful close handshake or a blocked writer.
func ConnectWebSocket(ctx context.Context, address string, options WebSocketOptions) (WebSocketLike, error) {
	if err := ctx.Err(); err != nil {
		return nil, err
	}
	maxMessage, maxHeader := options.MaxMessageBytes, options.MaxHeaderBytes
	if maxMessage == 0 {
		maxMessage = 4 * 1024 * 1024
	}
	if maxHeader == 0 {
		maxHeader = 64 * 1024
	}
	if maxMessage < 0 || maxHeader < 0 {
		return nil, errors.New("WebSocket byte limits must be positive")
	}
	target, err := url.Parse(address)
	if err != nil || target.Hostname() == "" || target.User != nil || target.Fragment != "" || target.Opaque != "" || (target.Scheme != "ws" && target.Scheme != "wss") || strings.ContainsAny(address, "\\\r\n\t ") {
		return nil, errors.New("WebSocket URL must be ws/wss without credentials or a fragment")
	}
	if target.Scheme == "ws" {
		target.Scheme = "http"
	} else {
		target.Scheme = "https"
	}
	header := make(http.Header)
	for name, values := range options.Header {
		if name == "" {
			return nil, errors.New("Invalid WebSocket request header")
		}
		for _, character := range name {
			if !(character >= 'a' && character <= 'z' || character >= 'A' && character <= 'Z' || character >= '0' && character <= '9' || strings.ContainsRune("!#$%&'*+-.^_`|~", character)) {
				return nil, errors.New("Invalid WebSocket request header")
			}
		}
		switch strings.ToLower(name) {
		case "host", "upgrade", "connection", "content-length", "transfer-encoding", "sec-websocket-key", "sec-websocket-version", "sec-websocket-protocol", "sec-websocket-extensions":
			return nil, errors.New("Invalid WebSocket request header")
		}
		for _, value := range values {
			for _, character := range value {
				if character < 32 || character > 126 {
					return nil, errors.New("Invalid WebSocket request header")
				}
			}
			header.Add(name, value)
		}
	}
	nonce := make([]byte, 16)
	if _, err := rand.Read(nonce); err != nil {
		return nil, err
	}
	key := base64.StdEncoding.EncodeToString(nonce)
	header.Set("Upgrade", "websocket")
	header.Set("Connection", "Upgrade")
	header.Set("Sec-WebSocket-Key", key)
	header.Set("Sec-WebSocket-Version", "13")
	ctx, cancel := context.WithCancel(ctx)
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, target.String(), nil)
	if err != nil {
		cancel()
		return nil, errors.New("Invalid WebSocket request URL")
	}
	request.Header = header
	transport := options.Transport
	if transport == nil {
		native := &http.Transport{MaxResponseHeaderBytes: maxHeader, DisableCompression: true, TLSNextProto: map[string]func(string, *tls.Conn) http.RoundTripper{}}
		defer native.CloseIdleConnections()
		transport = &http.Client{Transport: native, CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse }}
	}
	response, err := transport.Do(request)
	if err != nil {
		cancel()
		// net/http wraps errors with the request URL, which can contain API keys.
		var wrapped *url.Error
		if errors.As(err, &wrapped) {
			return nil, wrapped.Err
		}
		return nil, err
	}
	fail := func(message string) (WebSocketLike, error) {
		cancel()
		if response != nil && response.Body != nil {
			response.Body.Close()
		}
		return nil, errors.New(message)
	}
	if response == nil || response.Body == nil || response.StatusCode != http.StatusSwitchingProtocols {
		return fail("WebSocket handshake was not accepted")
	}
	upgrade := strings.ToLower(strings.TrimSpace(strings.Join(response.Header.Values("Upgrade"), ",")))
	connection := false
	for _, value := range response.Header.Values("Connection") {
		for _, token := range strings.Split(value, ",") {
			if strings.EqualFold(strings.TrimSpace(token), "upgrade") {
				connection = true
			}
		}
	}
	digest := sha1.Sum([]byte(key + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"))
	accepted := response.Header.Values("Sec-WebSocket-Accept")
	if upgrade != "websocket" || !connection || len(accepted) != 1 || strings.TrimSpace(accepted[0]) != base64.StdEncoding.EncodeToString(digest[:]) || len(response.Header.Values("Sec-WebSocket-Extensions")) != 0 || len(response.Header.Values("Sec-WebSocket-Protocol")) != 0 {
		return fail("Invalid WebSocket upgrade response")
	}
	body, ok := response.Body.(io.ReadWriteCloser)
	if !ok {
		return fail("WebSocket transport did not return a duplex body")
	}
	socket := &webSocket{body: body, ctx: ctx, cancel: cancel, maxMessage: maxMessage}
	socket.stop = context.AfterFunc(ctx, func() { socket.closeBody() })
	if err := ctx.Err(); err != nil {
		socket.Close()
		return nil, err
	}
	return socket, nil
}

type webSocket struct {
	body        io.ReadWriteCloser
	ctx         context.Context
	cancel      context.CancelFunc
	stop        func() bool
	maxMessage  int
	once        sync.Once
	closed      atomic.Bool
	closeError  error
	readMutex   sync.Mutex
	receiveDone bool
	writeMutex  sync.Mutex
}

func (s *webSocket) closeBody() error {
	s.once.Do(func() { s.closed.Store(true); s.closeError = s.body.Close() })
	return s.closeError
}
func (s *webSocket) Close() error { s.stop(); s.cancel(); return s.closeBody() }

func (s *webSocket) Send(ctx context.Context, message WebSocketMessage) error {
	var data []byte
	opcode := byte(1)
	switch value := message.(type) {
	case WebSocketText:
		data = []byte(value)
	case *WebSocketText:
		if value != nil {
			data = []byte(*value)
		} else {
			return errors.New("Invalid WebSocket message")
		}
	case WebSocketBinary:
		data = []byte(value)
		opcode = 2
	case *WebSocketBinary:
		if value != nil {
			data = []byte(*value)
			opcode = 2
		} else {
			return errors.New("Invalid WebSocket message")
		}
	default:
		return errors.New("Invalid WebSocket message")
	}
	if len(data) > s.maxMessage {
		return errors.New("WebSocket message exceeds byte limit")
	}
	if opcode == 1 && !utf8.Valid(data) {
		return errors.New("Invalid WebSocket UTF-8")
	}
	return s.sendFrame(ctx, opcode, data)
}
func (s *webSocket) sendFrame(ctx context.Context, opcode byte, data []byte) error {
	stop := context.AfterFunc(ctx, func() { s.Close() })
	defer stop()
	s.writeMutex.Lock()
	defer s.writeMutex.Unlock()
	if err := ctx.Err(); err != nil {
		s.Close()
		return err
	}
	if err := s.ctx.Err(); err != nil {
		return err
	}
	header := []byte{0x80 | opcode}
	switch {
	case len(data) < 126:
		header = append(header, 0x80|byte(len(data)))
	case len(data) < 65536:
		header = binary.BigEndian.AppendUint16(append(header, 0xfe), uint16(len(data)))
	default:
		header = binary.BigEndian.AppendUint64(append(header, 0xff), uint64(len(data)))
	}
	mask := make([]byte, 4)
	if _, err := rand.Read(mask); err != nil {
		return err
	}
	frame := make([]byte, len(header)+4+len(data))
	copy(frame, header)
	copy(frame[len(header):], mask)
	for index, value := range data {
		frame[len(header)+4+index] = value ^ mask[index%4]
	}
	n, err := s.body.Write(frame)
	if cause := ctx.Err(); cause != nil {
		s.Close()
		return cause
	}
	if cause := s.ctx.Err(); cause != nil {
		return cause
	}
	if err == nil && n != len(frame) {
		err = io.ErrShortWrite
	}
	if err != nil {
		s.Close()
	}
	return err
}
func (s *webSocket) Receive(ctx context.Context) (WebSocketMessage, error) {
	stop := context.AfterFunc(ctx, func() { s.Close() })
	defer stop()
	s.readMutex.Lock()
	defer s.readMutex.Unlock()
	if s.receiveDone {
		return nil, io.EOF
	}
	if err := ctx.Err(); err != nil {
		s.receiveDone = true
		s.Close()
		return nil, err
	}
	if err := s.ctx.Err(); err != nil {
		s.receiveDone = true
		s.Close()
		return nil, err
	}
	if s.closed.Load() {
		s.receiveDone = true
		return nil, io.EOF
	}
	message, err := s.receive(ctx)
	if cause := ctx.Err(); cause != nil {
		message, err = nil, cause
	} else if cause := s.ctx.Err(); cause != nil {
		message, err = nil, cause
	}
	if err != nil {
		s.receiveDone = true
		s.Close()
	}
	return message, err
}

func (s *webSocket) readFull(buffer []byte) error {
	_, err := io.ReadFull(s.body, buffer)
	if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
		return &WebSocketClosed{Code: 1006}
	}
	return err
}

func (s *webSocket) receive(ctx context.Context) (WebSocketMessage, error) {
	data := []byte{}
	messageOpcode := byte(0)
	for {
		var header [2]byte
		if err := s.readFull(header[:]); err != nil {
			return nil, err
		}
		final, opcode := header[0]&0x80 != 0, header[0]&15
		if header[0]&0x70 != 0 || header[1]&0x80 != 0 || (opcode > 2 && opcode != 8 && opcode != 9 && opcode != 10) {
			return nil, errors.New("Invalid WebSocket frame header")
		}
		length := uint64(header[1] & 127)
		if opcode >= 8 && (!final || length > 125) {
			return nil, errors.New("Invalid WebSocket control frame")
		}
		switch length {
		case 126:
			var extended [2]byte
			if err := s.readFull(extended[:]); err != nil {
				return nil, err
			}
			length = uint64(binary.BigEndian.Uint16(extended[:]))
			if length < 126 {
				return nil, errors.New("Nonminimal WebSocket frame length")
			}
		case 127:
			var extended [8]byte
			if err := s.readFull(extended[:]); err != nil {
				return nil, err
			}
			length = binary.BigEndian.Uint64(extended[:])
			if length < 65536 || length >= 1<<63 {
				return nil, errors.New("Invalid WebSocket frame length")
			}
		}
		if opcode < 8 {
			if (opcode == 0) != (messageOpcode != 0) {
				return nil, errors.New("Invalid WebSocket fragmentation")
			}
			if length > uint64(s.maxMessage-len(data)) {
				return nil, errors.New("WebSocket message exceeds byte limit")
			}
		}
		payload := make([]byte, int(length))
		if err := s.readFull(payload); err != nil {
			return nil, err
		}
		switch opcode {
		case 8:
			if len(payload) == 1 {
				return nil, errors.New("Invalid WebSocket close payload")
			}
			code := 1005
			reason := ""
			if len(payload) >= 2 {
				code = int(binary.BigEndian.Uint16(payload))
				reason = string(payload[2:])
				if !((code >= 1000 && code <= 1003) || (code >= 1007 && code <= 1014) || (code >= 3000 && code <= 4999)) {
					return nil, errors.New("Invalid WebSocket close code")
				}
				if !utf8.ValidString(reason) {
					return nil, errors.New("Invalid WebSocket UTF-8")
				}
			}
			if err := s.sendFrame(ctx, 8, payload); err != nil {
				return nil, err
			}
			if code == 1000 || code == 1005 {
				return nil, io.EOF
			}
			return nil, &WebSocketClosed{Code: code, Reason: reason}
		case 9:
			if err := s.sendFrame(ctx, 10, payload); err != nil {
				return nil, err
			}
			continue
		case 10:
			continue
		}
		if opcode != 0 {
			messageOpcode = opcode
		}
		data = append(data, payload...)
		if final {
			if messageOpcode == 2 {
				return WebSocketBinary(data), nil
			}
			if !utf8.Valid(data) {
				return nil, errors.New("Invalid WebSocket UTF-8")
			}
			return WebSocketText(data), nil
		}
	}
}
